### PR TITLE
Introduce UserSession::verifyAuthHeaders which validates the session …

### PR DESCRIPTION
**Pull request by DeepDiver1975**
…against given auth headers like basic auth or oauth

## Related Issue
https://github.com/owncloud/core/issues/28707

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by SamuAlfageme** [comment]
@DeepDiver1975 is this still a WIP?

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 is this still a WIP?

yes still WIP - will be back on this after the conference ...

**Comment by michaelstingl** [comment]
@PVince81 ping

**Comment by PVince81** [comment]
@DeepDiver1975 any chance to finish this for 10.0.4 ? It is a blocker for all clients. Without this, they all need to build in ugly workarounds.

**Comment by PVince81** [comment]
@DeepDiver1975 please post a list of remaining tasks/issues so someone else can take over

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 please post a list of remaining tasks/issues so someone else can take over

- [ ] test if this is already enough to fullfill the requirement
- [ ] make sure nothing else was broken

**Comment by PVince81** [comment]
> test if this is already enough to fullfill the requirement

@michaelstingl @SamuAlfageme

**Comment by michaelstingl** [comment]
@jesmrec @nasli @davigonz Please check with mobile platforms ^

**Comment by DeepDiver1975** [comment]
god damn - my latest changes of today are resulting in an infinite loop ..... 💥

**Comment by nasli** [comment]
👍  @DeepDiver1975 ping us again when is ready to test

**Comment by DeepDiver1975** [comment]
> 👍 @DeepDiver1975 ping us again when is ready to test

@nasli ping 😉

File: lib/private/App/AppManager.php [hunk 1]
```diff
@@ -124,10 +124,10 @@ public function getInstalledApps() {
 	/**
 	 * List all apps enabled for a user
 	 *
-	 * @param \OCP\IUser $user
+	 * @param \OCP\IUser|null $user
```
Comment (butonic) [inline lib/private/App/AppManager.php]: hm then checkAppForUser in line 165 also needs te allow null in the phpdocs

Comment (DeepDiver1975) [reply lib/private/App/AppManager.php]: indeed

**Comment by nasli** [comment]
Great!! Thanks @DeepDiver1975, spoiler! Jesus has made first tests and seems to work fine, we let you know

**Comment by jesmrec** [comment]
ping us when it is (finally) ready.

**Comment by PVince81** [comment]
so it seems the code is ready apart from the PHPDoc comment ?

this checkbox item still requires attention:

> make sure nothing else was broken

What needs to be retested ? Everything auth related, all auth types, shibboleth, etc ?

**Comment by PVince81** [comment]
@davitol is going to try and break this with different auth types.

@DeepDiver1975 assuming this is done apart from the PHPDoc issue

**Comment by davitol** [comment]
Tests results:

    ❌  basic auth on Webdav: It is not listed. It is needed a forwardport of https://github.com/owncloud/core/pull/28879
    ✅ web UI session
    ✅ auth with application token instead of password
    ✅ oauth2

**Comment by davitol** [comment]
✅ Public links and fed shares also went fine

**Comment by jesmrec** [comment]
will be this included in 10.0.4?

**Comment by PVince81** [comment]
considering that this is impairing clients I'd like to have this in 10.0.4 RC1

**Comment by davitol** [comment]
> onsidering that this is impairing clients I'd like to have this in 10.0.4 RC1

Then a backport PR is needed too

**Comment by phil-davis** [comment]
and [WIP] in title and "2-Developing" labels need to be removed so we know @DeepDiver1975 is happy.

**Comment by PVince81** [comment]
I talked with @DeepDiver1975 and he's going to finalize this PR, mostly related to the IServiceLoader bit. Once this is done we can backport for 10.0.4

**Comment by DeepDiver1975** [comment]
@PVince81 this is for 10.0.4 - right? I need to adjust the since annotation then

**Comment by PVince81** [comment]
@DeepDiver1975 yes please, thanks

File: lib/public/App/IServiceLoader.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ */
+
+
+namespace OCP\App;
+
+use OCP\IUser;
+
+/**
+ * Interface IServiceLoader
```
Comment (PVince81) [inline lib/public/App/IServiceLoader.php]: PHPDoc... what's a service loader, what's a service, etc.

File: lib/public/App/IServiceLoader.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+interface IServiceLoader {
+
+	/**
+	 * @param array $xmlPath
+	 * @param IUser|null $user
+	 * @return \Generator
+	 * @since 10.0.4
+	 */
+	public function load(array $xmlPath, IUser $user = null);
```
Comment (PVince81) [inline lib/public/App/IServiceLoader.php]: PHPDoc, what are we loading here ? People looking at the docs of \OC::$server::load will wonder

**Review by PVince81 — APPROVED** [review]
Please add missing PHPDoc then merge and backport 👍

**Comment by phil-davis** [comment]
Backport stable10 #29525

**Comment by PVince81** [comment]
Regression: https://github.com/owncloud/core/issues/30157

**Pull request merged**